git-hooks/commit-msg

```diff
@@ -0,0 +1,134 @@
+    ${SYNTAX}
+`
+
+const unknownType = (type) => `
+    ERROR: Unknown commit type!
+    ----------------------------------------
+    Given <type>:
+    ${type}
+
+    Expected <type> to be one of the following:
+        ${Object.keys(TYPES).join(', ')}
+
+    Syntax:
+    ${SYNTAX}
+`
+
+const TYPES = {
+    'build': 'update/modify build system and/or dependencies',
+    'ci': 'modification of CI configuration',
+    'docs': 'documentation-only changes',
+    'feat': 'add new functionality (semver MINOR)',
+    'fix': 'patch bug/error (semver PATCH)',
+    'perf': 'code changes that improve performance',
+    'refactor': 'code changes that neither fix a bug nor add a feature',
+    'style': 'code changes that do not modify algorithms (white space, linting, etc.)',
+    'test': 'add missing tests or update existing tests',
+}
+
+/**
+ * @param string header
+ * @returns object - object with the following keys:
+ *  - isBreaking: true if header implies a breaking change with an exclamation point
+ *  - scope: matched value for the optional <scope> portion of the header
+ *  - summary: matched value for the <summary> portion of the header
+ *  - type: matched value for the <type> portion of the header
+ */
+function parseHeader (header) {
+    let pattern = /^(?<type>[^\(!]+)(\((?<scope>.+)\))?(?<isBreaking>!)?: (?<summary>.+)$/
+
+    // check header for matches against the regex pattern
+    let matches = pattern.exec(header)
+
+    // default capture group values if there are no matches
+    let groups = {
+        isBreaking: null,
+        scope: null,
+        summary: null,
+        type: null,
+    }
+
+    // replace default capture group values with the real values
+    if (matches && matches.groups) {
+        groups = matches.groups
+    }
+
+    return {
+        ...groups,
+        isBreaking: !!groups.isBreaking,
+        matches,
+    }
+}
+
+/**
+ * Validate format of the message header
+ *
+ * @returns boolean
+ */
+function isValidHeader (header) {
+    let { matches, type } = parsed = parseHeader(header)
+
+    if (!matches) {
+        console.error(incorrectFormat(header))
+        return false
+    }
+
+    if (!TYPES.hasOwnProperty(type)) {
+        console.error(unknownType(type))
+        return false
+    }
+
+    return true
+}
+
+/**
+ * Splits raw commit message into an array of message lines (excluding comments)
+ * @param string message - commit message contents
+ * @returns array<string>
+ */
+function getMessageLines (message) {
+    let allLines = message.split('\n')
+
+    // filter out comments
+    let nonCommentLines = allLines.filter( line => !line.match(/^#/) )
+
+    return nonCommentLines
+}
+
+// define logic
+function main () {
+    let msgBuffer = fs.readFileSync(path.resolve(__dirname, '..', process.argv[2]))
+    let msgString = msgBuffer.toString()
+
+    let msgLines = getMessageLines(msgString)
+    let header = msgLines[0]
+
+    let isValid = isValidHeader(header)
```

Comment:
I could've placed the call to `isValidHeader` in the `if` condition below, but there's a good chance that the logic for validation may expand in the future, so I'm leaving it as a separate variable definition.